sycl/doc/extensions/proposed/sycl_ext_oneapi_async_memory_alloc.asciidoc

```diff
@@ -0,0 +1,729 @@
+
+void async_free(handler h, void *ptr);
+----
+
+_Effects_: Enqueues a commands to `q` or the SYCL queue associated with `h`
+which will asynchornously free the memory allocation at the address of `ptr`.
+Memory will be freed from the memory pool to be used by other asynchronous
+malloc commands which execute later, and will not free to the OS until the SYCL
+queue associated with the asycnhronous allocation command has been synchornized
+with.
+
+_Throws_: An exception with the `errc::invalid` error code if `ptr` is not the
+address of a memory allocation allocated to a memory pool.
+
+[_Note:_ Accessing the memory at the address of `ptr`after the asynchronous free
+command has completed execution is undefined behaviour. _{endnote}_]
+
+
+=== Memory pool lifetimes
+
+The lifetime of memory allocated via a memory pool is as follows.
+
+When an asynchornous malloc command is executed it will first look to the
+memory pool to opportunistically allocate memory already available there. The
+SYCL RT may also, when possible, look at all asynchronous malloc and free
+commands and their dependencies in order to more efficiently schedule memory
+allocations. If the memory pool does not have the requirement memory available
+to allocate for the asynchronous malloc command it will then look to allocate
+from the OS to the memory pool to provide the required memory.
+
+When an asynchornous free command is executed it will free up memory in the 
+memory pool to be used for a later asynchonrous malloc command. Memory is not
+freed back to the OS at this point.
+
+When a SYCL queue which enqueued asynchronous malloc commands is synchronized
+with, i.e. when `queue::wait` oo `queue::wait_and_throw` is called, the SYCL RT\
+will check the current total memory allocated after to the memory pool against
+the memory pool's deallocation threshold. If the total memory allocated islarger
+than than the threshold then the memory will be freed back to the OS, otherwise
+it will not.
+
+[_Note:_ An implementation should maintain memory allocated to a memory pool to
+the size specified by the deallocation threshold when possible, however, an
+implementation is permitted to release this memory back to the OS if it is
+required by another memory pool or an application in another process.
+_{endnote}_]
+
+
+== Implementation notes
+
+It is expected that for L0 this extension will be implemented within the L0
+adapter, by reserving allocations for the memory pool and opportunistically
+re-using the memory allocated based on the command lists being enqueued to the
+L0 driver.
+
+It is expected that for CUDA this extension will be implemented by mapping onto
+the CUDA stream-ordered allocator feature.
+
+
+== Issues
+
+. Should we allow mixing asynchornous and synchornous memory commands?
++
+--
+*UNRESOLVED*: CUDA allows memory allocated with the asynchornous malloc command
+to be freed with the regular sycnhronous free command, should we extend this
```

Comment:
```suggestion
*UNRESOLVED*: CUDA allows memory allocated with the asynchronous malloc command
to be freed with the regular synchronous free command, should we extend this
```

Reply:
Fixed in https://github.com/intel/llvm/pull/14800/commits/2c51adf28ece21c0e91c2c0ec85440ae1153e536.